Userland/Applications/PixelPaint/Tools/WandSelectTool.cpp

```diff
@@ -0,0 +1,186 @@
+
+    // Create Mask which will track already-colored pixels
+    Mask flood_mask = Mask::empty(bitmap.rect());
+    Mask selection_mask = Mask::empty(bitmap.rect());
+    Queue<Gfx::IntPoint> points_to_visit = Queue<Gfx::IntPoint>();
+
+    points_to_visit.enqueue({ start_position.x(), start_position.y() });
+    selection_mask.set(start_position.x(), start_position.y(), 0xFF);
+    flood_mask.set(start_position.x(), start_position.y(), 0xFF);
+
+    // This implements a non-recursive flood fill. This is a breadth-first search of paintable neighbors
+    // As we find neighbors that are paintable we update their pixel, add them to the queue, and mark them in the mask
+    while (!points_to_visit.is_empty()) {
+        auto current_point = points_to_visit.dequeue();
+        Gfx::IntPoint candidate_point = {};
+
+        if (current_point.x() > 0) {
+            candidate_point = current_point.moved_left(1);
+            if (flood_mask.get(candidate_point.x(), candidate_point.y()) == 0 && meets_wand_threshold(candidate_point, bitmap, target_color, threshold_normalized_squared)) {
+                points_to_visit.enqueue(candidate_point);
+                selection_mask.set(candidate_point.x(), candidate_point.y(), 0xFF);
+            }
+            flood_mask.set(candidate_point.x(), candidate_point.y(), 0xFF);
+        }
+
+        if (current_point.x() < bitmap.width() - 1) {
+            candidate_point = current_point.moved_right(1);
+            if (flood_mask.get(candidate_point.x(), candidate_point.y()) == 0 && meets_wand_threshold(candidate_point, bitmap, target_color, threshold_normalized_squared)) {
+                points_to_visit.enqueue(candidate_point);
+                selection_mask.set(candidate_point.x(), candidate_point.y(), 0xFF);
+            }
+            flood_mask.set(candidate_point.x(), candidate_point.y(), 0xFF);
+        }
+
+        if (current_point.y() > 0) {
+            candidate_point = current_point.moved_up(1);
+            if (flood_mask.get(candidate_point.x(), candidate_point.y()) == 0 && meets_wand_threshold(candidate_point, bitmap, target_color, threshold_normalized_squared)) {
+                points_to_visit.enqueue(candidate_point);
+                selection_mask.set(candidate_point.x(), candidate_point.y(), 0xFF);
+            }
+            flood_mask.set(candidate_point.x(), candidate_point.y(), 0xFF);
+        }
+
+        if (current_point.y() < bitmap.height() - 1) {
+            candidate_point = current_point.moved_down(1);
+            if (flood_mask.get(candidate_point.x(), candidate_point.y()) == 0 && meets_wand_threshold(candidate_point, bitmap, target_color, threshold_normalized_squared)) {
+                points_to_visit.enqueue(candidate_point);
+                selection_mask.set(candidate_point.x(), candidate_point.y(), 0xFF);
+            }
+            flood_mask.set(candidate_point.x(), candidate_point.y(), 0xFF);
+        }
```

Comment:
This looks fairly copy/paste duplicated, maybe replace it with a loop like:

```suggestion
auto candidate_points = Array {
    current_point.moved_left(1),
    current_point.moved_right(1),
    current_point.moved_up(1),
    current_point.moved_down(1)
};

for (auto candidate_point : candidate_points) {
    if (!bitmap.rect().contains(candidate_point))
        continue;
    if (flood_mask.get(candidate_point.x(), candidate_point.y()) == 0 && meets_wand_threshold(candidate_point, bitmap, target_color, threshold_normalized_squared)) {
        points_to_visit.enqueue(candidate_point);
        selection_mask.set(candidate_point.x(), candidate_point.y(), 0xFF);
    }
    flood_mask.set(candidate_point.x(), candidate_point.y(), 0xFF);
}
```

Reply:
Thanks for that. I did feel some degree of bad doing it that way, didn't know about Array{}. I've updated the code to match this suggestion. I'll also adopt this change over on the BucketTool PR